cli/src/commands/mod.rs

```diff
@@ -2351,6 +2430,134 @@ Please use `jj new 'all:x|y'` instead of `jj new --allow-large-revsets x y`.",
+    // If we're editing, start at the working-copy commit.
+    // Everything else starts from our direct parent.
+    let start_id = if edit {
+        current_wc_id
+    } else {
+        match current_wc.parent_ids() {
+            [parent_id] => parent_id,
+            _ => return Err(user_error("Cannot run `jj next` on a merge commit")),
+        }
+    };
+    let find_descendant_at = |amount: u64| -> Rc<RevsetExpression> {
+        RevsetExpression::commit(start_id.clone()).descendants_at(amount)
+    };
+    let target_expression = if edit {
+        find_descendant_at(amount)
+    } else {
+        find_descendant_at(amount).minus(&RevsetExpression::commit(current_wc_id.clone()))
+    };
+    let targets: Vec<Commit> = target_expression
+        .resolve(workspace_command.repo().as_ref())?
+        .evaluate(workspace_command.repo().as_ref())?
+        .iter()
+        .commits(workspace_command.repo().store())
+        .take(2)
+        .try_collect()?;
+    let target = match targets.as_slice() {
+        [target] => target,
+        [] => {
+            // We found no descendant.
+            return Err(user_error(format!(
+                "No descendant found {amount} commit{} forward",
+                if amount > 1 { "s" } else { "" }
+            )));
+        }
+        _ => {
+            // TODO(#NNN) We currently cannot deal with multiple children, which result
+            // from branches. Prompt the user for resolution.
+            return Err(user_error("Ambiguous target commit"));
+        }
+    };
+    let target_short = short_commit_hash(target.id());
+    // We're editing, just move to the target commit.
+    if edit {
+        // We're editing, the target must be rewritable.
+        workspace_command.check_rewritable(target)?;
+        let mut tx = workspace_command
+            .start_transaction(&format!("next: {current_short} -> editing {target_short}"));
+        tx.edit(target)?;
+        tx.finish(ui)?;
+        return Ok(());
+    }
+    let mut tx =
+        workspace_command.start_transaction(&format!("next: {current_short} -> {target_short}"));
+    // Move the working-copy commit to the new parent.
+    tx.check_out(target)?;
+    tx.finish(ui)?;
+    Ok(())
+}
+
+fn cmd_prev(ui: &mut Ui, command: &CommandHelper, args: &PrevArgs) -> Result<(), CommandError> {
+    let mut workspace_command = command.workspace_helper(ui)?;
+    let edit = args.edit;
+    let amount = args.amount;
+    let current_wc_id = workspace_command
+        .get_wc_commit_id()
+        .ok_or_else(|| user_error("This command requires a working copy"))?;
+    let current_wc = workspace_command.repo().store().get_commit(current_wc_id)?;
+    let current_short = short_commit_hash(current_wc.id());
+    let start_id = if edit {
+        current_wc_id
+    } else {
+        match current_wc.parent_ids() {
+            [parent_id] => parent_id,
+            _ => return Err(user_error("Cannot run `jj prev` on a branching parent")),
```

Comment:
nit: I think we typically call it a "merge commit" (as you did in the `jj next` implementation)